database/migrations/2021_06_09_202500_refactor_photo_model.php

```diff
@@ -0,0 +1,479 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Config;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Schema;
+use Illuminate\Support\Facades\Storage;
+
+class RefactorPhotoModel extends Migration
+{
+	const THUMBNAIL_DIM = 200;
+	const THUMBNAIL2X_DIM = 400;
+
+	const VARIANT_ORIGINAL = 0;
+	const VARIANT_MEDIUM2X = 1;
+	const VARIANT_MEDIUM = 2;
+	const VARIANT_SMALL2X = 3;
+	const VARIANT_SMALL = 4;
+	const VARIANT_THUMB2X = 5;
+	const VARIANT_THUMB = 6;
+
+	/**
+	 * Maps a size variant to the path prefix (directory) where the file for that size variant is stored.
+	 * Use this array to avoid the anti-pattern "magic constants" throughout the whole code.
+	 */
+	const VARIANT_2_PATH_PREFIX = [
+		'big',
+		'medium',
+		'medium',
+		'small',
+		'small',
+		'thumb',
+		'thumb',
+	];
+
+	const VALID_VIDEO_TYPES = [
+		'video/mp4',
+		'video/mpeg',
+		'image/x-tga', // mpg; will be corrected by the metadata extractor
+		'video/ogg',
+		'video/webm',
+		'video/quicktime',
+		'video/x-ms-asf', // wmv file
+		'video/x-ms-wmv', // wmv file
+		'video/x-msvideo', // Avi
+		'video/x-m4v', // Avi
+		'application/octet-stream', // Some mp4 files; will be corrected by the metadata extractor
+	];
+
+	const VARIANT_2_WIDTH_ATTRIBUTE = [
+		'width',
+		'medium2x_width',
+		'medium_width',
+		'small2x_width',
+		'small_width',
+	];
+
+	const VARIANT_2_HEIGHT_ATTRIBUTE = [
+		'height',
+		'medium2x_height',
+		'medium_height',
+		'small2x_height',
+		'small_height',
+	];
+
+	/**
+	 * Run the migrations.
+	 *
+	 * @return void
+	 */
+	public function up()
+	{
+		$this->upgradeInconsistentAttributes();
+		// We are brutal here and simply drop the old table, as the symlinks
+		// will be re-created on-the-fly when required
+		Schema::dropIfExists('sym_links');
+		Schema::dropIfExists('size_variants');
```

Comment:
Yes, we do. I ran into this problem. If the first run of the upgrade fails (for some reason) and you run the upgrade a second time after the reason for the failed first run has been fixed, the table is already there and must be dropped.

If such an error occurs luckily no data is lost, because the whole data migration is in a transaction.

Reply:
Yes, we got several such error reports from users after the 4.3.4 release, where the migration clearly failed for some reason but when they rerun it they get unrelated error messages because the new columns already exist. Perhaps we should institute a policy that adding new columns or tables in a migration should not cause failures if they already exist...